### PR TITLE
[SPARK-30147][SQL] Trim the string when cast string type to booleans

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Now, we trim the string when casting string value to those `canCast` types values, e.g. int, double, decimal, interval, date, timestamps, except for boolean. 
This behavior makes type cast and coercion inconsistency in Spark.
Not fitting ANSI SQL standard either.
```
If TD is boolean, then
Case:
a) If SD is character string, then SV is replaced by
    TRIM ( BOTH ' ' FROM VE )
    Case:
    i) If the rules for literal in Subclause 5.3, “literal”, can be applied to SV to determine a valid
value of the data type TD, then let TV be that value.
   ii) Otherwise, an exception condition is raised: data exception — invalid character value for cast.
b) If SD is boolean, then TV is SV
```
In this pull request, we trim all the whitespaces from both ends of the string before converting it to a bool value. This behavior is as same as others, but a bit different from sql standard, which trim only spaces. 

### Why are the changes needed?

Type cast/coercion consistency

### Does this PR introduce any user-facing change?

yes, string with whitespaces in both ends will be trimmed before converted to booleans.

e.g. `select cast('\t true' as boolean)` results `true` now, before this pr it's `null`
### How was this patch tested?

add unit tests

### CODE REVIEW

**Comment by yaooqinn** [comment]
@cloud-fan @maropu, thanks for reviewing.

File: docs/sql-migration-guide.md [hunk 1]
```diff
@@ -222,10 +222,12 @@ license: |
...
 
-  - Since Spark 3.0, when casting string value to integral types, including tinyint, smallint, int and bigint type, the leading and trailing white spaces(<= ACSII 32) will be trimmed before convert to integral values, e.g. `cast(' 1 ' as int)` results `1`. In Spark version 2.4 and earlier, the result will be `null`.
+  - Since Spark 3.0, when casting string value to integral types, including tinyint, smallint, int and bigint type, the leading and trailing whitespaces(<= ACSII 32) will be trimmed before convert to integral values, e.g. `cast(' 1 ' as int)` results `1`. In Spark version 2.4 and earlier, the result will be `null`.
+
+  - Since Spark 3.0, when casting string value to date, timestamp and interval values, the leading and trailing whitespaces(<= ACSII 32) will be trimmed before casing, e.g. `cast('2019-10-10\t as date)` results the date value `2019-10-10`. In Spark version 2.4 and earlier, only the trailing space will be removed, thus, the result is `null`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md]: can we merge these notes? It's the same behavior change but for different data types.

Comment (yaooqinn) [reply docs/sql-migration-guide.md]: sgtm

**Review by cloud-fan — APPROVED** [review]

**Comment by yaooqinn** [comment]
Seems we get a flaky test failure here.

**Review by maropu — APPROVED** [review]
Looks nice, thanks, @yaooqinn !

**Comment by maropu** [comment]
Thanks, merged to master.

**Pull request closed without merging**